iib/common/tracing.py

```diff
@@ -0,0 +1,211 @@
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+"""Configures the Global Tracer Provider and exports the traces to the OpenTelemetry Collector.
+
+The OpenTelemetry Collector is configured to receive traces via OTLP over HTTP.
+The OTLP exporter is configured to use the environment variables
+OTEL_EXPORTER_OTLP_TRACES_ENDPOINT and OTEL_EXPORTER_OTLP_TRACES_HEADERS to
+configure the endpoint and headers for the OTLP exporter.
+The OTLP* environment variables are configured in the docker-compose.yaml
+and podman-compose.yaml files for iib workers and api.
+
+Usage:
+    @instrument_tracing()
+    def func():
+        pass
+
+    @instrument_tracing()
+    class MyClass:
+        def func1():
+            pass
+        def _func2():
+            pass
+
+"""
+import os
+import functools
+import inspect
+import logging
+from typing import Dict
+from opentelemetry import trace
+from opentelemetry.trace import Tracer
+from opentelemetry.trace import SpanKind, Status, StatusCode
+from opentelemetry.sdk.resources import Resource, SERVICE_NAME
+from opentelemetry.sdk.trace import TracerProvider
+from opentelemetry.sdk.trace.export import (
+    BatchSpanProcessor,
+)
+from opentelemetry.propagate import set_global_textmap
+from opentelemetry.trace.propagation import (
+    set_span_in_context,
+)
+from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
+from opentelemetry.trace.propagation.tracecontext import TraceContextTextMapPropagator
+
+
+log = logging.getLogger(__name__)
+propagator = TraceContextTextMapPropagator()
+
+
+class TracingWrapper:
+    """Wrapper class that will wrap all methods of a calls with the instrument_tracing decorator."""
+
+    __instance = None
+
+    def __new__(cls):
+        """Create a new instance if one does not exist."""
+        if TracingWrapper.__instance is None:
+            log.info('Creating TracingWrapper instance')
+            cls.__instance = super().__new__(cls)
+            otlp_exporter = OTLPSpanExporter(
+                endpoint=f"{os.getenv('OTEL_EXPORTER_OTLP_ENDPOINT')}/v1/traces",
+            )
+            cls.provider = TracerProvider(
+                resource=Resource.create({SERVICE_NAME: 'iib-auto-manual'})
```

Comment:
Should the service name be `iib-auto-manual` or the value passed by the env variable OTEL_RESOURCE_ATTRIBUTES or OTEL_SERVICE_NAME ?

Reply:
No this was my local testing value that got pushed. Updated it to take it from the `OTEL_SERVICE_NAME`